File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    List<List<Integer>> ans = new ArrayList<>();
    public void helper(int index, int target, int[] candidates, List<Integer> curr) {
        if (target == 0) {
            ans.add(new ArrayList<>(curr));
            return;
        }
        if (index == candidates.length) return;
        if (target >= candidates[index]) {
            curr.add(candidates[index]);
            helper(index, target - candidates[index], candidates, curr);
            curr.remove(curr.size() - 1);
        }
        helper(index + 1, target, candidates, curr);
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        helper(0, target, candidates, new ArrayList<>());
        return ans;
    }
}

File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    vector<vector<int>> ans;
    void helper(int index, vector<int>& cdts, vector<int>& curr, int t) {
        if (t == 0) {
            ans.push_back(curr);
            return;
        }
        if (index == cdts.size()) return;
        if (t >= cdts[index]) {
            curr.push_back(cdts[index]);
            helper(index, cdts, curr, t-cdts[index]);
            curr.pop_back();
        }
        helper(index+1, cdts, curr, t);
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> curr;
        helper(0, candidates, curr, target);
        return ans;
    }
};